costmap_cspace/include/costmap_cspace/costmap_3d_layer/footprint.h

```diff
@@ -202,112 +219,116 @@ class Costmap3dLayerFootprint : public Costmap3dLayerBase
         const int gy = std::lround((i / msg->info.width) * resolution_scale) + oy;
         if (static_cast<size_t>(gx) >= map->info.width ||
             static_cast<size_t>(gy) >= map->info.height)
           continue;
-        const int8_t val = msg->data[i];
-        if (val < 0)
-        {
-          continue;
-        }
-        else if (val == 0)
-        {
-          int8_t& m = map->getCost(gx, gy, yaw);
-          if (m < 0)
-            m = 0;
-          continue;
-        }
+        // If the cell is unknown in parent map and also updated map,
+        // the cell is never measured.
+        // Never measured cells should be marked unknown
+        // to be correctly processed in the planner.
+        unknown_buf_[i] = msg->data[i] < 0 && map->getCost(gx, gy, yaw) < 0;
+      }
+    }
+    for (size_t i = 0; i < msg->data.size(); i++)
+    {
+      const int gx = std::lround((i % msg->info.width) * resolution_scale) + ox;
+      const int gy = std::lround((i / msg->info.width) * resolution_scale) + oy;
+      if (static_cast<size_t>(gx) >= map->info.width ||
+          static_cast<size_t>(gy) >= map->info.height)
+        continue;
+      const int8_t val = msg->data[i];
+      if (val < 0)
+      {
+        continue;
+      }
+      else if (val == 0)
+      {
+        int8_t& m = map->getCost(gx, gy, yaw);
+        if (m < 0)
+          m = 0;
+        continue;
+      }
 
-        const int map_x_min = std::max(gx - range_max_, 0);
-        const int map_x_max = std::min(gx + range_max_, static_cast<int>(map->info.width) - 1);
-        const int map_y_min = std::max(gy - range_max_, 0);
-        const int map_y_max = std::min(gy + range_max_, static_cast<int>(map->info.height) - 1);
-        for (int map_y = map_y_min; map_y <= map_y_max; ++map_y)
+      const int map_x_min = std::max(gx - range_max_, 0);
+      const int map_x_max = std::min(gx + range_max_, static_cast<int>(map->info.width) - 1);
+      const int map_y_min = std::max(gy - range_max_, 0);
+      const int map_y_max = std::min(gy + range_max_, static_cast<int>(map->info.height) - 1);
+      for (int map_y = map_y_min; map_y <= map_y_max; ++map_y)
+      {
+        // Use raw pointers for faster iteration
+        int8_t* cost_addr = &(map->getCost(map_x_min, map_y, yaw));
+        const char* cs_addr = &(cs_template_.e(map_x_min - gx, map_y - gy, yaw));
+        for (int n = 0; n <= map_x_max - map_x_min; ++n, ++cost_addr, ++cs_addr)
         {
-          // Use raw pointers for faster iteration
-          int8_t* cost_addr = &(map->getCost(map_x_min, map_y, yaw));
-          const char* cs_addr = &(cs_template_.e(map_x_min - gx, map_y - gy, yaw));
-          for (int n = 0; n <= map_x_max - map_x_min; ++n, ++cost_addr, ++cs_addr)
-          {
-            const int8_t c = *cs_addr * val / 100;
-            if (c > 0 && *cost_addr < c)
-              *cost_addr = c;
-          }
+          const int8_t c = *cs_addr * val / 100;
```

Comment:
We could divide `val` by `100` outside of the double for loop.

Reply:
As it's integer, it will drop digits.

Reply:
Yes, if `val / 100` is stored as a float, floating point multiplication and cast are performed here and the calculation cost will be increased.